src/Upgrade/Infrastructure/VersionControlSystem/Generator/ViolationBodyMessageBuilder.php

```diff
@@ -9,121 +9,61 @@
 {
     /**
-     * @param array<\Upgrade\Application\Dto\ViolationDto> $violations
-     * @param array<\Upgrade\Domain\Entity\Package> $packageDtos
-     *
-     * @return string
+     * @var \Upgrade\Infrastructure\VersionControlSystem\Generator\CheckerViolationMessageBuilderRegistry
      */
-    public function buildViolationsTable(array $violations, array $packageDtos): string
-    {
-        $text = '';
-
-        foreach ($this->getGroupedViolationsByMessage($violations) as $message => $groupedViolationsByMessage) {
-            $text .= PHP_EOL . '[WARNING] ' . $message . PHP_EOL;
-            $text .= '| Package | Release | Project file | '
-                . PHP_EOL
-                . '|---------|---------|--------------|'
-                . PHP_EOL;
-
-            $text .= $this->getMessageViolationsTableLines($groupedViolationsByMessage, $packageDtos);
-            $text .= PHP_EOL;
-        }
+    protected CheckerViolationMessageBuilderRegistry $checkerViolationMessageBuilderRegistry;
 
-        return $text;
+    /**
+     * @param \Upgrade\Infrastructure\VersionControlSystem\Generator\CheckerViolationMessageBuilderRegistry $checkerViolationMessageBuilderRegistry
+     */
+    public function __construct(
+        CheckerViolationMessageBuilderRegistry $checkerViolationMessageBuilderRegistry
+    ) {
+        $this->checkerViolationMessageBuilderRegistry = $checkerViolationMessageBuilderRegistry;
     }
 
     /**
-     * @param array<\Upgrade\Application\Dto\ViolationDto> $groupedViolationsByMessage
+     * @param array<\Upgrade\Application\Dto\ViolationDtoInterface> $violations
      * @param array<\Upgrade\Domain\Entity\Package> $packageDtos
      *
      * @return string
      */
-    protected function getMessageViolationsTableLines(array $groupedViolationsByMessage, array $packageDtos): string
+    public function buildViolationsMessage(array $violations, array $packageDtos): string
     {
         $text = '';
+        $groupedViolations = $this->groupViolationsByType($violations);
 
-        foreach ($this->getGroupedViolationsByPackage($groupedViolationsByMessage) as $package => $violationsByPackage) {
-            $text .= implode(
-                ' | ',
-                [
-                    '**' . $package . '**',
-                    sprintf($this->getPackageReleaseLink($packageDtos, $package)),
-                    implode(
-                        '<br>',
-                        array_unique(array_map(static fn (ViolationDto $violations): string => $violations->getTarget(), $violationsByPackage)),
-                    ),
-                    PHP_EOL,
-                ],
-            );
+        foreach ($groupedViolations as $type => $violations) {
+            $checkerMessageBuilder = $this->checkerViolationMessageBuilderRegistry->getBuilderByType($type);
+            $text .= $checkerMessageBuilder->buildViolationsMessage($violations, $packageDtos);
+            $text .= PHP_EOL;
```

Comment:
`$checkerMessageBuilder->buildViolationsMessage()` already has PHP_EOL at the end. Are you sure this is necessary here?

Reply:
Yeap. We need additional empty line here to separate dynamic evaluator block